### ISSUE TITLE
Graphic artefacts in PC/GEOS with a lot of VESA modes (mouse not erased completely)

### ISSUE THREAD

**Issue by bolle732** [stsp/dosemu2#301]
I'll rise an 'old' problem of PC/GEOS under DOSEmu which happens in most of the VESA modes. This problem exist since the 1.x days of DOSEmu. Under DOSBox or on real hardware, the problem does not occur.

If I move the mouse from top to bottom, the mouse cursor is not erased correctly all the time. It leaves the top most line of the cursor on the screen as an artefact. When I move up the mouse and pass by this artefacts, they get removed.

![mouse-glitches](https://cloud.githubusercontent.com/assets/8956508/22916989/6c7ece66-f282-11e6-8005-fab4a61a7e34.png)

Further interesting, the problem only seems to exist on the right side of the screen and not on the whole.

There is no clear indication of a systematic problem depending on resolution or colour depth. For example (x, y, bpp):
```
GOOD: 640x480x8, 800x600x8, 1024x768x16
BAD : 640x480x[15|16|24], 800x600x[15|16|24], 1024x768x[8|15|24]
```
I've checked the mouse code with the help of putting debug statements into it. I ended up in `gcursor.c`, function `erase_graphics_cursor`. But surprise, this code get never executed. A quick research told me, that in VBE, no standard for a hardware cursor exist. Therefore, I believe that the program is drawing the mouse cursor itself.

Then I tried to check the VGAEMU / VBE code. But there, I got a little lost. From the debug log, I see that we should be in 'bank' mode and not in 'lfb' mode.
```
VGAEmu: vga_emu_fault: write access to bank region, address 0xa0000, page 0xa0, vga page 0x0
VGAEmu: vga_emu_fault: cs:eip = 41d8:14fa, instr: f3 ab 2b fa 2b fa 4d 74 14 03 fe 72 08 2e 39 3e
```

Initialization has been this:
```
INT10: near fc00:00000113
INT10: AH=4f AL=00
VBE: [0x00] vbe_info: es:di = 0x45c9:0x0000
VBE: [0x00] vbe_info: return value 0
INT10: near fc00:00000113
INT10: AH=4f AL=01
VBE: [0x01] vbe_mode_info: cx = 0x0112, es:di = 0x45c9:0x0000
VBE: [0x01] vbe_mode_info: return value 0
INT10: near fc00:00000113
INT10: AH=4f AL=02
VBE: [0x02] vbe_set_mode: bx = 0x0112
INT10: set_video_mode: mode 0x112
VGAEmu: vga_emu_setmode: requested mode: 0x112 (80 x 30)
VGAEmu: vga_emu_setmode: mode found in first run
VGAEmu: vga_emu_setmode: mode 0x112, (640 x 480 x 24, 80 x 30, 8 x 16, 64k at 0xa000)
VGAEmu: vga_emu_setmode: scan_len = 1920
VGAEmu: vga_emu_setmode: setting up components...
VGAEmu: DAC_init done
VGAEmu: Attr_init done
VGAEmu: Seq: Setting font offsets to 0x0000/0x0000 (0x00)
VGAEmu: Seq: Seq_init done
VGAEmu: CRTC_init done
VGAEmu: GFX_init done
VGAEmu: Misc_init done
VGAEmu: Herc_init done
VGAEmu: vgaemu_adj_cfg: seq.addr_mode = chain4 (dword)
VGAEmu: vgaemu_adj_cfg: horizontal_total = 100
VGAEmu: vgaemu_adj_cfg: horizontal_retrace_start = 84
VGAEmu: vgaemu_adj_cfg: horizontal_retrace_end = 0
VGAEmu: vgaemu_adj_cfg: horizontal_blanking_start = 80
VGAEmu: vgaemu_adj_cfg: horizontal_blanking_end = 2
VGAEmu: vgaemu_adj_cfg: horizontal_display_end = 80
VGAEmu: vgaemu_adj_cfg: multiplier = 8
VGAEmu: vgaemu_adj_cfg: width = 640
VGAEmu: vgaemu_adj_cfg: vertical_total = 523
VGAEmu: vgaemu_adj_cfg: vertical_retrace_start = 490
VGAEmu: vgaemu_adj_cfg: vertical_retrace_end = 12
VGAEmu: vgaemu_adj_cfg: vertical_blanking_start = 479
VGAEmu: vgaemu_adj_cfg: vertical_blanking_end = 4
VGAEmu: vgaemu_adj_cfg: vertical_display_end = 479
VGAEmu: vgaemu_adj_cfg: vertical_multiplier = 1
VGAEmu: vgaemu_adj_cfg: height = 480
VGAEmu: vgaemu_adj_cfg: line_compare = 1023
VGAEmu: vga_emu_setmode: mode initialized
INT10: mapped cursor: start 14, end 15
scanlines=d1f9 480
VBE: [0x02] vbe_set_mode: return value 0
modify_mode: geometry changed to 640 x 480, scan_len = 1920 bytes
vgaemu: set page 0 dirty
vgaemu: set page 1 dirty
vgaemu: set page 2 dirty
vgaemu: set page 3 dirty
vgaemu: set page 4 dirty
vgaemu: set page 5 dirty
VBE: [0x01] vbe_mode_info: return value 0
```
What I've seen, are a lot of `vgaemu: set page <#> dirty`. But I think this shouldn't be a reason for the problem.

Should I upload the `boot.log` (`-D9+v`) ? Can this be tracked down ?

**Comment by bolle732** [comment]
I've done a little more tests. This got my attention:
```
VBE: [0x05.0] vbe_display_window::set: bl = 0x00, dx = 0x0009
VBE: [0x05.0] vbe_display_window: accessed via direct call
VBE: [0x05.0] vbe_display_window::set: return value 0
VBE: [0x05.0] vbe_display_window::set: return value 0
VBE: [0x05.0] vbe_display_window::set: bl = 0x00, dx = 0x0007
VBE: [0x05.0] vbe_display_window: accessed via direct call
VBE: [0x05.0] vbe_display_window::set: return value 0
VBE: [0x05.0] vbe_display_window::set: bl = 0x00, dx = 0x0008
VBE: [0x05.0] vbe_display_window: accessed via direct call
VBE: [0x05.0] vbe_display_window::set: return value 0
```
How can the `set: return value 0` come twice in a row ?

**Comment by stsp** [comment]
Does this depend on $_cpu_vm and $_cpu_emu settings?
It would also be very interesting if you try Bart's branches
that are currently sitting in PRs. Though it is most likely just
some artifact of vga_emu_update code.

**Comment by stsp** [comment]
> How can the set: return value 0 come twice in a row ?

Have you stripped the _unrelated_ reports in between,
or is it virgin? VGA code in main thread runs in parallel
with the renderer thread, and if renderer was outputting
something in between, the lines could be lost.
We need to protect the logging with mutex.
As you get this reproducible, please send the patch. :)

**Comment by bolle732** [comment]
Ok, need some time to do the tests with vm / emu settings and the braches from Bart. Will report about the findings.

The output of `set: return value 0` is a `grep` from the `boot.log`. I can upload the zipped 3MB (uncompressed 70MB) if needed.

**Comment by stsp** [comment]
What kind of grep? If you grepped for that
line then no wonder.

**Comment by bolle732** [comment]
I've used a simple `tail -f boot.log | grep 'vbe_display_window'`. As I read the code of the function `vbe_display_window`, I expect for any `set: return value X` a preceding `set: bl = XX, dx = XX`. Or I'm completely wrong ?

So, I'm not even sure, if this has anything to do with the artefacts.

I've tested different branches from Bart ('kvm_log_dirty', 'simx86_e_vgaemu' and 'kvm_guest_paging'). They all behave the same way.

Further, I've tested with '$_cpu_vm' ('vm86', 'full') and '$_cpu_emu' ('vm86', 'kvm'). Didn't found any combination where the artefacts don't appear.

**Comment by stsp** [comment]
> Or I'm completely wrong ?

I think I answered this already.

> So, I'm not even sure, if this has anything to do with the artefacts.

It doesn't.

> Didn't found any combination where the artefacts don't appear.

This may be a problem with vga update code then.
Try adding dirty_all_video_pages() into render.c:452
after sem_wait().

**Comment by bolle732** [comment]
Yes, with the `dirty_all_video_pages()` in 'render.c:452', the artefacts has been disappeared. On the downside, the CPU usage increased about 10 times.

This test went fine. Where should I look next ?

**Comment by stsp** [comment]
> This test went fine. Where should I look next ?

vga_emu_update()

**Comment by stsp** [comment]
Or it is possible to be a problem with remapper.
In this case perhaps you can see if with 'dosemu -X'
some filter settings affect that.

**Comment by bolle732** [comment]
Ok, I've made a PR which is needed to be able to debug the VGAEMU updates.

The code of the VGAEMU is a little bit complicated. Would be nice to clean up the obsolete part... Worth a week of work :(

Regarding my problem, I can fix the artefacts with the following change in 'vgaemu.c' / '__vga_emu_update()':

```
-  veut->update_start = i << 12;
+  veut->update_start = i << 12 - 1;
```

In my eyes, this is a quick fix / workaround. Any comments / ideas ?

**Comment by bolle732** [comment]
I'll do more tests too.

**Comment by stsp** [comment]
Something like this was expected, but the
interesting finding in what you did is that you
need to add/subtract only one byte, and not
one page as I thought you need. This pretty
much makes me think that the problem is indeed
in the remapper, as vgaemu update code operates
with pages, not with bytes.
So the next area of your investigations may be
the 'dosemu -X' vs filters test, and looking into
_remap_remap_mem() function. But I don't think
one can quickly parse that function. :( I'll probably
need to look into that problem myself.

I fully agree with you that vgaemu code is junk
and is asking for being completely replaced with
something from qemu. :) But its not a simple task.
Just writing a new renderer code for dosemu2
took me a month of work (the new renderer was
sponsored so I could throw a lot of work into it).
Then I was planning to rewrite a remapper and
then - replace vgaemu. Why I think the remapper
should be rewritten first, is because it needs to
became stateless, then the parallel rendering can
be done. Also it needs to have the API that is advanced
enough to easily handle the things like mem_wrap
and planar modes. Then the vgaemu code would be
much, much simpler.
But now I find these tasks unfeasible.

**Comment by stsp** [comment]
I cleaned up vga_emu_update() for you.
It appears all the messy logic is simply not
used by the new renderer, so it was not
difficult to remove the redundant bits.
Should be tested a bit.

**Comment by bolle732** [comment]
Thanks a lot.

Could you please remove the undefined 'start_page' variable too ? This is used by 'vga_deb_update' in 'vgaemu.c:1840'.

**Comment by bolle732** [comment]
> Something like this was expected, but the
interesting finding in what you did is that you
need to add/subtract only one byte, and not
one page as I thought you need. This pretty
much makes me think that the problem is indeed
in the remapper, as vgaemu update code operates
with pages, not with bytes.

Doesn't come the + before >> ?

So that `veut->update_start = i << 12 - 1;` is equal to `veut->update_start = i << 11;` ?

**Comment by stsp** [comment]
> Could you please remove

Done.

> Doesn't come the + before >> ?

Ah but then your fix is completely wrong.
Please see what minimal value you need to
add/subtract. Without knowing the minimum,
nothing can be proposed.

**Comment by bolle732** [comment]
This drives me a little bit crazy.

There is no single value I can subtract or add to the 'veut->update_*'. With extending by page size in both directions, the artefacts disappears mostly, but still occurs sometimes. I think, this is due to how the updates are tracked in linear RAM. I would prefer a 2 dimensional way of the dirty bitmap...

So, I'll start again with observations from the beginning. Step by step, this could go longer then I like...

Checking the log with 1280x1024x24 and 'DEBUG_UPDATE' enabled I see:
```
modify_mode: geometry changed to 1280 x 1024, scan_len = 3840 bytes
VGAEmu: vga_emu_update: display = 0 (page = 0) - 3928320 (page = 959), update_pos = 0
VGAEmu: dirty_map[0 - 1024k] (0 = RO/NONE, 1 = RW = dirty)
   1111.1111.1111.1111 1111.1111.1111.1111 1111.1111.1111.1111 1111.1111.1111.1111
   1111.1111.1111.1111 1111.1111.1111.1111 1111.1111.1111.1111 1111.1111.1111.1111
   1111.1111.1111.1111 1111.1111.1111.1111 1111.1111.1111.1111 1111.1111.1111.1111
   1111.1111.1111.1111 1111.1111.1111.1111 1111.1111.1111.1111 1111.1111.1111.1111
vgaemu: set page 0 dirty
```
Shouldn't the 'display_end' value be 3932160 (1280 x 1024 x 3 bytes) instead of 3928320 ? Is this by intention or is it off by one 'scan_len' (3840) ?

Two improvement here would be to extend 'print_dirty_map()' to support more then 1024k and add the value of dirty in the output of "set page X (VALUE)".

Regarding logging, would it make sense to include the underlines of the functions who has them in the name (like '__vga_emu_update' instead of only 'vga_emu_update') ?

**Comment by stsp** [comment]
> I would prefer a 2 dimensional way of the dirty bitmap...

The bitmap tracks dirty memory pages. I am not sure
what kind of 2-dimentional it can be.

> Shouldn't the 'display_end' value be 3932160 (1280 x 1024 x 3 bytes) instead of 3928320 ?

This may depend on vga.line_compare or vga.mem.wrap
values, see update_graphics_screen(). I have simplified
this function a lot yesterday and of course I could over-simplify
it and add the bug...

> Regarding logging, would it make sense to include the underlines

It didn't have underlines initially but now it does.
I don't think that matters much.
If you want, you can use __func__ of course.

**Comment by bolle732** [comment]
> The bitmap tracks dirty memory pages. I am not sure what kind of 2-dimentional it can be.

Sorry, wrong direction of me.

I spent a lot of time trying to understand the code. But I didn't come to an end regarding the artefacts. Shame on me...

I believe, there is a lot of needed improvement. For example, I've patched the logging of `_vgaemu_dirty_page()` like this:
```
diff --git a/src/base/dev/vga/vgaemu.c b/src/base/dev/vga/vgaemu.c
index 389ac95..29b7a6d 100644
--- a/src/base/dev/vga/vgaemu.c
+++ b/src/base/dev/vga/vgaemu.c
@@ -2565,7 +2565,7 @@ static void _vgaemu_dirty_page(int page, int dirty)
     dosemu_error("vgaemu: page out of range, %i (%i)\n", page, vga.mem.pages);
     return;
   }
-  v_printf("vgaemu: set page %i dirty\n", page);
+  v_printf("_vgaemu_dirty_page: set page %i dirty (%i -> %i)\n", page, vga.mem.dirty_map[page], dirty);
   /* prot_mtx should be locked by caller */
   vga.mem.dirty_map[page] = dirty;
```
Using `-D9+v` and only starting up to the command prompt of FreeDOS and exiting, I get more the 17000 lines like this:
`_vgaemu_dirty_page: set page X dirty (1 -> 1)`

For X=1 alone, there are around 200 consecutive lines doing the same.

**Comment by stsp** [comment]
Do you see this is something like CGA, or
in LFB modes? In case of an LFB modes I
would highly suspect the bug, just as you do.
I can take a look into artifacts myself when
time permits, but it would be nice if you can
isolate the interesting bugs like this one.

**Comment by bolle732** [comment]
I haven't tested this specifically regarding CGA / LFB. Only started DOSEMU with text mode. As far as I can interpret the source, it could be due to the initialization of the video memory. Need to check `vga_mark_dirty()`...

**Comment by bolle732** [comment]
Putting debug outputs into `vga_mark_dirty()` gives me around 3500 calls only starting FreeDOS into 'command.com':
```
vgaemu: set page 0 dirty (1 -> 1)
vga_mark_dirty: length is 1
vga_mark_dirty: addr is 753664
vgaemu: set page 0 dirty (1 -> 1)
vga_mark_dirty: length is 1
vga_mark_dirty: addr is 753664
vgaemu: set page 0 dirty (1 -> 1)
vga_mark_dirty: length is 1
vga_mark_dirty: addr is 753664
vgaemu: set page 0 dirty (1 -> 1)
vga_mark_dirty: length is 1
vga_mark_dirty: addr is 753664
vgaemu: set page 0 dirty (1 -> 1)
vga_mark_dirty: length is 1
vga_mark_dirty: addr is 753664
vgaemu: set page 0 dirty (1 -> 1)
```

The `vga_mark_dirty()` is called by all `vga_write*` functions and some memory copy functions. Each call to `vga_mark_dirty()` produces a call of `vga_emu_adjust_protection(vga_page, 0, VGA_PROT_RW, 1);`.

Don't know what the proper fix would be. But it seems to me not very efficient... Especially if always the same address is in use ?!?

**Comment by bolle732** [comment]
Further, it seems to me that the functions `vga_emu_XXX_[un]lock()` and `pthread_mutex_[un]lock(&XXX_mtx)` are a little bit inconsistently used.

**Comment by stsp** [comment]
> I haven't tested this specifically regarding CGA / LFB. Only started DOSEMU with text mode.

Ah, for text mode it is likely fine.
It should not happen only in LFB and most VGA modes.

> call to vga_mark_dirty() produces a call of
> vga_emu_adjust_protection(vga_page, 0, VGA_PROT_RW, 1);.

But it should do nothing in case of the same
protection, as it has the "protection cache".

> pthread_mutex_[un]lock(&XXX_mtx) are a little bit inconsistently used.

What do you mean?

**Comment by bolle732** [comment]
>> call to vga_mark_dirty() produces a call of vga_emu_adjust_protection(vga_page, 0, VGA_PROT_RW, 1);.

> But it should do nothing in case of the same protection, as it has the "protection cache".

Ok, logically it isn't a problem. But in case of debugging it makes it difficult due to the big noise... Does it emulate every pixel in text mode ?

>> pthread_mutex_[un]lock(&XXX_mtx) are a little bit inconsistently used.

> What do you mean?

That I don't understand why `pthread_mutex_*` is called sometimes directly and sometimes the wrapper function `vga_emu_XXX_*` is used.

**Comment by stsp** [comment]
> But in case of debugging it makes it difficult due to the big noise...

Perhaps you can add the vga.mem.dirty_map check
into vga_mark_dirty() before calling _vga_emu_adjust_protection()?

> That I don't understand why pthread_mutex_* is called sometimes directly
> and sometimes the wrapper

It is called directly by the vgaemu code itself,
and the wrapper is for the external API.

**Comment by bolle732** [comment]
Yes, with putting an if before the `_vga_emu_adjust_protection()` in `vga_mark_dirty()` silence the noise.
```
  if (!vga.mem.dirty_map[vga_page])
    _vga_emu_adjust_protection(vga_page, 0, VGA_PROT_RW, 1);
```

**Comment by bolle732** [comment]
Regarding the artefacts, I have a new finding. The mouse cursor is 16x16 pixel. If I move the mouse by 1 pixel, this should result in 16 calls of vga_emu_fault(). When checking with 1024x768x16, I see this 16 calls and there are no artefacts. Under 1024x768x24, I've sometimes only 12 - 14 calls and the mouse artefacts appears.

Does this make sense ?

**Comment by bolle732** [comment]
Regarding the artefacts, I have a new finding. The mouse cursor is 16x16 pixel. If I move the mouse by 1 pixel down, I expect to see 16 calls of vga_emu_fault(). Checking this with 1024x768x16, I see 16 calls and there are no artefacts. Under 1024x768x24, I've sometimes only 12 - 14 calls and the mouse artefacts appears.

Does this make sense ? How can I debug this further ?

I'm looking at messages like this:
```
VBE: [0x05.0] vbe_display_window::set: return value 0
VGAEmu: vga_emu_fault: write access to bank region, address 0xa0000, page 0xa0, vga page 0x10
VGAEmu: vga_emu_fault: cs:eip = 41d8:2418, instr: 26 88 25 47 26 88 05 2e 8b 3e ff 3c 49 cb 2e 83
_vgaemu_dirty_page: set page 16 dirty (0 -> 1)
VGAEmu: vga_emu_fault: write access to bank region, address 0xa1000, page 0xa1, vga page 0x11
VGAEmu: vga_emu_fault: cs:eip = 41d8:145e, instr: ab 03 fa e2 f6 5f 59 4d 74 14 03 fe 72 08 2e 39
_vgaemu_dirty_page: set page 17 dirty (0 -> 1)
VGAEmu: vga_emu_fault: write access to bank region, address 0xa2000, page 0xa2, vga page 0x12
VGAEmu: vga_emu_fault: cs:eip = 41d8:145b, instr: aa 86 c3 ab 03 fa e2 f6 5f 59 4d 74 14 03 fe 72
_vgaemu_dirty_page: set page 18 dirty (0 -> 1)
VGAEmu: vga_emu_fault: write access to bank region, address 0xa3000, page 0xa3, vga page 0x13
VGAEmu: vga_emu_fault: cs:eip = 41d8:145e, instr: ab 03 fa e2 f6 5f 59 4d 74 14 03 fe 72 08 2e 39
```

**Comment by bolle732** [comment]
Another excerpt with artefacts:
```
INT10: near fc00:00000113
INT10: AH=4f AL=05
VBE: [0x05.0] vbe_display_window::set: bl = 0x00, dx = 0x0015
VBE: [0x05.0] vbe_display_window: accessed via direct call
_vgaemu_dirty_page: set page 336 dirty (0 -> 0)
_vgaemu_dirty_page: set page 337 dirty (0 -> 0)
_vgaemu_dirty_page: set page 338 dirty (0 -> 0)
_vgaemu_dirty_page: set page 339 dirty (0 -> 0)
_vgaemu_dirty_page: set page 340 dirty (0 -> 0)
_vgaemu_dirty_page: set page 341 dirty (0 -> 0)
_vgaemu_dirty_page: set page 342 dirty (0 -> 0)
_vgaemu_dirty_page: set page 343 dirty (0 -> 0)
_vgaemu_dirty_page: set page 344 dirty (0 -> 0)
_vgaemu_dirty_page: set page 345 dirty (0 -> 0)
_vgaemu_dirty_page: set page 346 dirty (0 -> 0)
_vgaemu_dirty_page: set page 347 dirty (0 -> 0)
_vgaemu_dirty_page: set page 348 dirty (0 -> 0)
_vgaemu_dirty_page: set page 349 dirty (0 -> 0)
_vgaemu_dirty_page: set page 350 dirty (0 -> 0)
_vgaemu_dirty_page: set page 351 dirty (0 -> 0)
VBE: [0x05.0] vbe_display_window::set: return value 0
kvm_vm86: entered
VGAEmu: vga_emu_fault: write access to bank region, address 0xa1725, page 0xa1, vga page 0x151
VGAEmu: vga_emu_fault: cs:eip = 41d8:2c69, instr: 26 89 05 2e 8a 44 02 26 88 45 02 83 c6 03 2e 03
_vgaemu_dirty_page: set page 337 dirty (0 -> 1)
kvm_vm86: entered
VGAEmu: vga_emu_fault: write access to bank region, address 0xa2325, page 0xa2, vga page 0x152
VGAEmu: vga_emu_fault: cs:eip = 41d8:2c69, instr: 26 89 05 2e 8a 44 02 26 88 45 02 83 c6 03 2e 03
_vgaemu_dirty_page: set page 338 dirty (0 -> 1)
kvm_vm86: entered
VGAEmu: vga_emu_fault: write access to bank region, address 0xa3b25, page 0xa3, vga page 0x153
VGAEmu: vga_emu_fault: cs:eip = 41d8:2c69, instr: 26 89 05 2e 8a 44 02 26 88 45 02 83 c6 03 2e 03
_vgaemu_dirty_page: set page 339 dirty (0 -> 1)
kvm_vm86: entered
VGAEmu: vga_emu_fault: write access to bank region, address 0xa4725, page 0xa4, vga page 0x154
VGAEmu: vga_emu_fault: cs:eip = 41d8:2c69, instr: 26 89 05 2e 8a 44 02 26 88 45 02 83 c6 03 2e 03
_vgaemu_dirty_page: set page 340 dirty (0 -> 1)
kvm_vm86: entered
VGAEmu: vga_emu_fault: write access to bank region, address 0xa5325, page 0xa5, vga page 0x155
VGAEmu: vga_emu_fault: cs:eip = 41d8:2c69, instr: 26 89 05 2e 8a 44 02 26 88 45 02 83 c6 03 2e 03
_vgaemu_dirty_page: set page 341 dirty (0 -> 1)
kvm_vm86: entered
VGAEmu: vga_emu_fault: write access to bank region, address 0xa6b25, page 0xa6, vga page 0x156
VGAEmu: vga_emu_fault: cs:eip = 41d8:2c69, instr: 26 89 05 2e 8a 44 02 26 88 45 02 83 c6 03 2e 03
_vgaemu_dirty_page: set page 342 dirty (0 -> 1)
kvm_vm86: entered
VGAEmu: vga_emu_fault: write access to bank region, address 0xa7725, page 0xa7, vga page 0x157
VGAEmu: vga_emu_fault: cs:eip = 41d8:2c69, instr: 26 89 05 2e 8a 44 02 26 88 45 02 83 c6 03 2e 03
_vgaemu_dirty_page: set page 343 dirty (0 -> 1)
kvm_vm86: entered
VGAEmu: vga_emu_fault: write access to bank region, address 0xa8325, page 0xa8, vga page 0x158
VGAEmu: vga_emu_fault: cs:eip = 41d8:2c69, instr: 26 89 05 2e 8a 44 02 26 88 45 02 83 c6 03 2e 03
_vgaemu_dirty_page: set page 344 dirty (0 -> 1)
kvm_vm86: entered
VGAEmu: vga_emu_fault: write access to bank region, address 0xa9b25, page 0xa9, vga page 0x159
VGAEmu: vga_emu_fault: cs:eip = 41d8:2c69, instr: 26 89 05 2e 8a 44 02 26 88 45 02 83 c6 03 2e 03
_vgaemu_dirty_page: set page 345 dirty (0 -> 1)
kvm_vm86: entered
VGAEmu: vga_emu_fault: write access to bank region, address 0xaa725, page 0xaa, vga page 0x15a
VGAEmu: vga_emu_fault: cs:eip = 41d8:2c69, instr: 26 89 05 2e 8a 44 02 26 88 45 02 83 c6 03 2e 03
_vgaemu_dirty_page: set page 346 dirty (0 -> 1)
kvm_vm86: entered
VGAEmu: vga_emu_fault: write access to bank region, address 0xab325, page 0xab, vga page 0x15b
VGAEmu: vga_emu_fault: cs:eip = 41d8:2c69, instr: 26 89 05 2e 8a 44 02 26 88 45 02 83 c6 03 2e 03
_vgaemu_dirty_page: set page 347 dirty (0 -> 1)
kvm_vm86: entered
VGAEmu: vga_emu_fault: write access to bank region, address 0xacb25, page 0xac, vga page 0x15c
VGAEmu: vga_emu_fault: cs:eip = 41d8:2c69, instr: 26 89 05 2e 8a 44 02 26 88 45 02 83 c6 03 2e 03
_vgaemu_dirty_page: set page 348 dirty (0 -> 1)
kvm_vm86: entered
INT10: near fc00:00000113
INT10: AH=4f AL=05
VBE: [0x05.0] vbe_display_window::set: bl = 0x00, dx = 0x0015
VBE: [0x05.0] vbe_display_window: accessed via direct call
_vgaemu_dirty_page: set page 336 dirty (0 -> 0)
_vgaemu_dirty_page: set page 349 dirty (0 -> 0)
_vgaemu_dirty_page: set page 350 dirty (0 -> 0)
_vgaemu_dirty_page: set page 351 dirty (0 -> 0)
VBE: [0x05.0] vbe_display_window::set: return value 0
kvm_vm86: entered
INT10: near fc00:00000113
INT10: AH=4f AL=05
VBE: [0x05.0] vbe_display_window::set: bl = 0x00, dx = 0x0015
VBE: [0x05.0] vbe_display_window: accessed via direct call
_vgaemu_dirty_page: set page 336 dirty (0 -> 0)
_vgaemu_dirty_page: set page 349 dirty (0 -> 0)
_vgaemu_dirty_page: set page 350 dirty (0 -> 0)
_vgaemu_dirty_page: set page 351 dirty (0 -> 0)
VBE: [0x05.0] vbe_display_window::set: return value 0
kvm_vm86: entered
VGAEmu: vga_emu_fault: write access to bank region, address 0xad725, page 0xad, vga page 0x15d
VGAEmu: vga_emu_fault: cs:eip = 41d8:29d8, instr: ab 46 46 43 43 8a 05 2e 88 07 2e 22 04 2e 32 84
_vgaemu_dirty_page: set page 349 dirty (0 -> 1)
kvm_vm86: entered
kvm_vm86: entered
VGAEmu: vga_emu_update: display = 0 (page = 0) - 8388608 (page = 2047), update_pos = 0
VGAEmu: dirty_map[0 - 1024k] (0 = RO/NONE, 1 = RW = dirty)
   0000.0000.0000.0000 0000.0000.0000.0000 0000.0000.0000.0000 0000.0000.0000.0000
   0000.0000.0000.0000 0000.0000.0000.0000 0000.0000.0000.0000 0000.0000.0000.0000
   0000.0000.0000.0000 0000.0000.0000.0000 0000.0000.0000.0000 0000.0000.0000.0000
   0000.0000.0000.0000 0000.0000.0000.0000 0000.0000.0000.0000 0000.0000.0000.0000
```
Could it be, that the INT10 has not terminated and another one interrupted the code path?

**Comment by stsp** [comment]
> Does this make sense ?

Partially.
You can't expect any predictable count of
vga_emu_fault()s because on every fault we
mark the page dirty and unprotect it. Then the
rendering thread scans for dirty pages, redraws
them and write-protects again.
It is possible that we have some race condition
so that your app dirties the page while the render
thread redraws it. Then render marks it as clean,
but it is actually not.

**Comment by bolle732** [comment]
That's what I see as the difference between the working resolution and the ones with artefacts.

Then, how can I debug this further ?

**Comment by stsp** [comment]
> Could it be, that the INT10 has not terminated and another one interrupted the code path?

I have told you already that we simply need
the mutex to protect logging from corruption
like this.

**Comment by stsp** [comment]
> Then, how can I debug this further ?

I added RENDER_THREADED define which you
can set to 0 to get non-threaded rendering.
This way we can see if you hit some race or not.
Also would be nice if you benchmark the non-threaded
mode and compare it in both X and SDL.

**Comment by stsp** [comment]
Interestingly, I see almost no improvement
with threaded rendering now, unless the
full-screen mode is used with filters and w/o
hw accel and with cpu-emu instead of kvm.
With such affairs there is probably no need
to work on a massively parallel rendering.
Its quite surprising that problems are fixing
themselves these days.

**Comment by bolle732** [comment]
Sorry, had been busy with a private case... I've did the bench right now.

PC/GEOS with 1024x768x16 @ 86c5d52de7559a728e2fba7055e16dfb7e9be553

RENDER_THREADED 1:

X:
13500 max itrs, CPU max 99%, CPU min (idle) <2%, CPU max (mouse) 10%

$_SDL_swrend = (on):
9300 max itrs, CPU max 97%, CPU min (idle) 2%, CPU max (mouse) 21%

$_SDL_swrend = (off):
11800 max itrs, CPU max 113%, CPU min (idle) 2%, CPU max (mouse) 25%


RENDER_THREADED 0:

X:
13100 max itrs, CPU max 104%, CPU min (idle) <2%, CPU max (mouse) 8%

$_SDL_swrend = (on):
8000 max itrs, CPU max 87%, CPU min (idle) <2%, CPU max (mouse) 24%

$_SDL_swrend = (off):
10600 max itrs, CPU max 98%, CPU min (idle) <2%, CPU max (mouse) 18%

Moving the mouse around under X doesn't redraws it well, even if I don't have artefacts in this resolution / colour depth. SDL redraws the mouse very vell, while under X it flickers.

**Comment by stsp** [comment]
> SDL redraws the mouse very vell, while under X it flickers.

Is this a regression?

**Comment by stsp** [comment]
> RENDER_THREADED 0:

This was not for the bench.
This was to isolate the possibility of the
race condition in the update logic.
The commit in another ticket was for
the bench. But its nice to know threaded
rendering is still slightly better.

**Comment by bolle732** [comment]
Regarding artefacts, there is no difference between RENDER_THREADED 0 or 1.

**Comment by stsp** [comment]
Oh. That's a bad news. No race condition.
Then this may be the remapper problem after
all. The easiest check I can think of, is to try
another remapper. You can do so by altering
SDL_csd to match the P16 color space.
SDL will start remapping the color space
internally then. It will be very slow, but good
for testing.
Btw, what kind of P16 do we have exactly?
Something like SDL_PIXELFORMAT_RGB565
I suppose.

**Comment by bolle732** [comment]
Sorry, haven't been able to get more time on this right now. First to sort out some private things.

In the meantime, I've found the 'vtext.exe' from Don Lewis (search for 'VESA24_2.ZIP'). With this VESA test, I can reproduce the behaviour of artefacts in some VESA modes too. For example, try mode '117' (good) and '118' (bad).

**Comment by stsp** [comment]
Git it from here:
http://cd.textfiles.com/simtel/simtel20/MSDOS/GRAPHICS/
but there are no mouse tests at all.
It only draws some gfx.

**Comment by bolle732** [comment]
Yes, there is no mouse test. I get the impression, that the problem is not purely related to the mouse. Only that the mouse will trigger it more easily.

With all the testing, I've seen sometimes missing redraws of some parts of (scan?) lines. As I mostly worked with 1024x768x64k due to the artefacts with the mouse, I didn't recognised it before.

So, I've now the impression, that the problem with high resolution VESA modes is somewhere deeper. As I couldn't isolate the root cause currently, I'm looking around to find other programs having a problem which could be related.

**Comment by stsp** [comment]
```

diff --git a/src/base/video/remap.c b/src/base/video/remap.c
index b9d1dd1..cc7d47a 100644
--- a/src/base/video/remap.c
+++ b/src/base/video/remap.c
@@ -1461,6 +1461,7 @@ void gen_4to16_all(RemapObject *);
 void gen_4to24_all(RemapObject *);
 void gen_4to32_all(RemapObject *);
 
+void gen_16to16_0(RemapObject *);
 void gen_16to16_1(RemapObject *);
 void gen_24to24_1(RemapObject *);
 void gen_32to32_1(RemapObject *);
@@ -1634,13 +1635,21 @@ static RemapFuncDesc remap_gen_list[] = {
   ),
 
   REMAP_DESC(
+    RFF_SCALE_1,
+    MODE_TRUE_16,
+    MODE_TRUE_16,
+    gen_16to16_0,
+    NULL
+  ),
+/*
+  REMAP_DESC(
     RFF_SCALE_1  | RFF_REMAP_LINES,
     MODE_TRUE_16,
     MODE_TRUE_16,
     gen_16to16_1,
     NULL
   ),
-
+*/
   REMAP_DESC(
     RFF_SCALE_1  | RFF_REMAP_LINES,
     MODE_TRUE_24,
@@ -2576,6 +2585,23 @@ void gen_4to32_all(RemapObject *ro)
  * 16 bit true color --> 16 bit true color
  * 15 bit true color --> 15 bit true color
  */
+void gen_16to16_0(RemapObject *ro)
+{
+  int i;
+  const unsigned char *src;
+  unsigned char *dst;
+
+  src = ro->src_image;
+  dst = ro->dst_image;
+
+error("remap %i\n", ro->src_height);
+  for(i = 0; i < ro->src_height; i++) {
+    memcpy(dst, src, ro->src_width << 1);
+    src += ro->src_scan_len;
+    dst += ro->dst_scan_len;
+  }
+}
+
 void gen_16to16_1(RemapObject *ro)
 {
   int i;
diff --git a/src/plugin/sdl/sdl.c b/src/plugin/sdl/sdl.c
index 6ad29a9..0ab6689 100644
--- a/src/plugin/sdl/sdl.c
+++ b/src/plugin/sdl/sdl.c
@@ -275,10 +275,10 @@ int SDL_init(void)
   }
 
   pixel_format = SDL_GetWindowPixelFormat(window);
-  if (pixel_format == SDL_PIXELFORMAT_UNKNOWN) {
-    error("SDL: unable to get pixel format\n");
-    pixel_format = SDL_PIXELFORMAT_RGB888;
-  }
+//  if (pixel_format == SDL_PIXELFORMAT_UNKNOWN) {
+//    error("SDL: unable to get pixel format\n");
+    pixel_format = SDL_PIXELFORMAT_RGB565;
+//  }
   SDL_PixelFormatEnumToMasks(pixel_format, &bpp, &rm, &gm, &bm, &am);
   SDL_csd.bits = bpp;
   SDL_csd.r_mask = rm;
```

**Comment by stsp** [comment]
The above patch is a test of the different
remapper. This is what I was suggesting
_you_ to do. :) In any case, it doesn't
remove the problem, which is very unexpected!
My test remapper remaps the entire screen -
I made it specially so that it is very close to
the test with `dirty_all_video_pages()` which
does remove the problem.
So while still no fix, I am sure this gives us
a very big clue. We only need to answer one
simple question: how does this remapper test
differ from the `dirty_all_video_pages()` test?
So far I have no idea, but at least it is a vector
for doing the further searchings.
Maybe this has something to do with the bank-switching.
GEOS does a lot of bank-switching, and in
the log it seems the bank-switches somehow
correlate with the glitch...

**Comment by stsp** [comment]
The below one fixes the problem:
```
diff --git a/src/base/video/remap.c b/src/base/video/remap.c
index b9d1dd1..227f3a3 100644
--- a/src/base/video/remap.c
+++ b/src/base/video/remap.c
@@ -1047,18 +1047,18 @@ static RectArea remap_mem_1(RemapObject *ro, int offset, int len)
     }
   }
   else {
-    ro->src_y0 = i1;
-    ro->src_y1 = j1;
-    if(j2) ro->src_y1++;
-    ro->dst_y0 = bre_d_0(i1, ro->src_height, ro->dst_height);
-    ro->dst_y1 = bre_d_0(j1, ro->src_height, ro->dst_height);
-    ra.y = ro->dst_y0;
-    ra.height = ro->dst_y1 - ro->dst_y0;
-    ro->src_offset = ro->dst_offset = 0;
     ro->src_x0 = ro->dst_x0 = ro->src_y0 = ro->dst_y0 = 0;
-    ro->src_x1 = ro->src_width; ro->dst_x1 = ro->dst_width;
-    ro->src_y1 = ro->src_height; ro->dst_y1 = ro->dst_height;
-    ra.height = ro->dst_height;
+    ro->src_y1 = ro->src_height;
+    ro->dst_y1 = ro->dst_height;
+    ro->src_offset = ro->dst_offset = 0;
+    ro->src_x1 = ro->src_width;
+    ro->dst_x1 = ro->dst_width;
+    ro->src_y1 = ro->src_height;
+    ro->dst_y1 = ro->dst_height;
+    ra.x = ro->dst_x0;
+    ra.y = ro->dst_y0;
+    ra.width = ro->dst_x1 - ra.x;
+    ra.height = ro->dst_y1 - ra.y;
     REMAP_AREA_DEBUG_FUNC(ro);
     ro->remap_func(ro);
   }
@@ -1461,6 +1461,7 @@ void gen_4to16_all(RemapObject *);
 void gen_4to24_all(RemapObject *);
 void gen_4to32_all(RemapObject *);
 
+void gen_16to16_0(RemapObject *);
 void gen_16to16_1(RemapObject *);
 void gen_24to24_1(RemapObject *);
 void gen_32to32_1(RemapObject *);
@@ -1634,13 +1635,21 @@ static RemapFuncDesc remap_gen_list[] = {
   ),
 
   REMAP_DESC(
+    RFF_SCALE_1,
+    MODE_TRUE_16,
+    MODE_TRUE_16,
+    gen_16to16_0,
+    NULL
+  ),
+/*
+  REMAP_DESC(
     RFF_SCALE_1  | RFF_REMAP_LINES,
     MODE_TRUE_16,
     MODE_TRUE_16,
     gen_16to16_1,
     NULL
   ),
-
+*/
   REMAP_DESC(
     RFF_SCALE_1  | RFF_REMAP_LINES,
     MODE_TRUE_24,
@@ -2576,6 +2585,22 @@ void gen_4to32_all(RemapObject *ro)
  * 16 bit true color --> 16 bit true color
  * 15 bit true color --> 15 bit true color
  */
+void gen_16to16_0(RemapObject *ro)
+{
+  int i;
+  const unsigned char *src;
+  unsigned char *dst;
+
+  src = ro->src_image;
+  dst = ro->dst_image;
+
+  for(i = 0; i < ro->src_height; i++) {
+    memcpy(dst, src, ro->src_width << 1);
+    src += ro->src_scan_len;
+    dst += ro->dst_scan_len;
+  }
+}
+
 void gen_16to16_1(RemapObject *ro)
 {
   int i;
diff --git a/src/plugin/sdl/sdl.c b/src/plugin/sdl/sdl.c
index 6ad29a9..0ab6689 100644
--- a/src/plugin/sdl/sdl.c
+++ b/src/plugin/sdl/sdl.c
@@ -275,10 +275,10 @@ int SDL_init(void)
   }
 
   pixel_format = SDL_GetWindowPixelFormat(window);
-  if (pixel_format == SDL_PIXELFORMAT_UNKNOWN) {
-    error("SDL: unable to get pixel format\n");
-    pixel_format = SDL_PIXELFORMAT_RGB888;
-  }
+//  if (pixel_format == SDL_PIXELFORMAT_UNKNOWN) {
+//    error("SDL: unable to get pixel format\n");
+    pixel_format = SDL_PIXELFORMAT_RGB565;
+//  }
   SDL_PixelFormatEnumToMasks(pixel_format, &bpp, &rm, &gm, &bm, &am);
   SDL_csd.bits = bpp;
   SDL_csd.r_mask = rm;
```

**Comment by stsp** [comment]
So somehow the problem appeared 2-fold.
When I installed the dummy remapper, I've
hit the bug in the dummy remapper handler,
that surprisingly gives the same glitches as
the original bug with RFF_REMAP_RECT.
That's why I thought the remapper swap
didn't change anything. Very confusing
combination of bugs.

**Comment by bolle732** [comment]
I've tested the patched remapper and it seems to work mostly in all 64k colour modes.

Under X, the artefacts still happens. Does X use the remapper the same way or not ?

**Comment by bolle732** [comment]
One more info: I'm running X with 24bit colours. Should there be an explicit remapper for 16to24_1 or 16to24_all ?

**Comment by stsp** [comment]
> Does X use the remapper the same way or not ?

As you can see, my test patch alters the SDL code
to use my test remapper. Since it doesn't touch the
X code, X keeps using RFF_REMAP_RECT that has
the original bug with exactly the same visual artifacts
as the one I fixed for the simple remapper...

> One more info: I'm running X with 24bit colours. Should there be an
> explicit remapper for 16to24_1 or 16to24_all ?

X uses _all while SDL uses _1 variants.
I wonder if in the future we can use a very simple
remappers, leaving the color space conversion to
SDL.

**Comment by bolle732** [comment]
Sorry, have been very busy with a private case. But now, back to this issue.

The above patches don't makes me thinking right. At one side, it's only for the SDL path, on the other, it increase the CPU load a lot. With SDL and moving around only the mouse, the CPU usage increased from ~30% to ~80%.

After a lot of debugging, I came up with the following patch:
```
diff --git a/src/base/video/remap.c b/src/base/video/remap.c
index b9d1dd1..3948fb6 100644
--- a/src/base/video/remap.c
+++ b/src/base/video/remap.c
@@ -954,6 +954,10 @@ static RectArea remap_mem_1(RemapObject *ro, int offset, int len)
   j1 = (offset + len) / ro->src_scan_len;
   j2 = (offset + len) % ro->src_scan_len;
 
+  // XXX: Temporary fix of artefacts (off by one ???)
+  if (i1 > 0) i1--;
+  if (j1 < ro->src_height) j1++;
+
   /* make sure it's all visible */
   if(i2 >= ro->src_width) i1++, i2 = 0, offset = i1 * ro->src_scan_len;
   if(i1 >= ro->src_height || i1 > j1) return ra;
```
With this patch, the artefacts disappeared in my tests with SDL and X under different high resolutions (up to 2048x1536) and colour depths (16bpp / 24bpp). The increase of the CPU usage is very little.

Currently, I'm thinking there is something like an off by one problem. Debugging the real root cause would require a lot more time.

Would a PR with this acceptable ?

**Comment by bolle732** [comment]
Regarding the above VTEST.EXE, I get the impression that it triggers another problem which is unrelated to the artefacts here and need to be investigated separately.

**Comment by stsp** [comment]
Yes, the patch was just to show where the
problem is (and it actually shows on a different
problem anyway).
I think to fix the real bug you should start somewhere
at remap.c:974 and change `if (i2)` to something like
`if (i2 && (i1 == j1))`. Same treatment should go to line
1013. And maybe there are more bugs in this function.

Don't worry, if you can't fix this, I'll make sure it is fixed
before tagging the next pre. Just didn't have enough
time for this. The problem is localized and it won't escape.
Of course if you fix it yourself, that would be welcome.

**Comment by bolle732** [comment]
Thanks for your comment. I would appreciate your help definitively.

I forgot to mention, that the related code is the one of the 'else if' for 'RFF_REMAP_LINES'  starting at line 1031.

I can see the code path with 'RFF_REMAP_RECT' when starting / terminating DOSEmu2 itself. But running PC/GEOS goes the 'RFF_REMAP_LINES' way.

**Comment by stsp** [comment]
Hmm, interesting. In fact, `RFF_REMAP_LINES` looks
like the most bug-free case here, and is quite simple too.
Can you try to replace at line 1040 `j1` with `ro->src_y1`?

**Comment by bolle732** [comment]
On line 1037, we have `ro->src_y1 = j1;`. So, unfortunately, I this doesn't change anything...

**Comment by stsp** [comment]
> So, unfortunately, I this doesn't change anything...

I can't parse this.
You _think_ this doesn't change, or you _tested_
this doesn't change? Obviously a missing word.

**Comment by bolle732** [comment]
This are the lines 1037 to 1040 of 'remap.c':
```
    ro->src_y1 = j1;
    if(j2) ro->src_y1++;
    ro->dst_y0 = bre_d_0(i1, ro->src_height, ro->dst_height);
    ro->dst_y1 = bre_d_0(j1, ro->src_height, ro->dst_height);
```
Sorry, my wrong that I've overlooked the line 1038. Ashes on my head.

But I've tested your advice too and it doesn't resolve the artefacts. Still trying to understand the code better.

**Comment by bolle732** [comment]
For me, the code is still very suspicious.

First, we have local variables holding something like the coordinates by scan length:
```
  i1 = offset / ro->src_scan_len;
  i2 = offset % ro->src_scan_len;
  j1 = (offset + len) / ro->src_scan_len;
  j2 = (offset + len) % ro->src_scan_len;
```
Note between: the names would be more friendly something like x0, y0, x1 and y1.
```
  /* make sure it's all visible */
  if(i2 >= ro->src_width) i1++, i2 = 0, offset = i1 * ro->src_scan_len;
  if(i1 >= ro->src_height || i1 > j1) return ra;
  if(j2 >= ro->src_width) j1++, j2 = 0;
  if(j1 >= ro->src_height) j1 = ro->src_height, j2 = 0;
```
Then we do checks with the coordinates by scan length together with pixel height and width. Further the special case with `i2 = 0`...

Sorry, but I currently don't get the whole picture behind that :(

For example with:
```
  if(i2 >= ro->src_width) i1++, i2 = 0, offset = i1 * ro->src_scan_len;
```
Is this an aligning to a multiply of the scan length ? But why does it increase i1 instead of decreasing it ? Shouldn't the aligned value be smaller so that we have the whole range for updating for sure ?

**Comment by stsp** [comment]
Sorry, ignore prev comment. They are not
sanity checks indeed.

**Comment by stsp** [comment]
That code looks fine to me.
All it does is to make sure things are inside
the src_width bounds. Can you try, for example,
instead doing `i1--` just leave `i1` untouched and
only write zero to i2?

**Comment by bolle732** [comment]
Yes, I tough that decreasing is not the last solution when it could be 0...

With leaving i1 untouched the artefacts still disappeared and it seems to work as expected:
```
diff --git a/src/base/video/remap.c b/src/base/video/remap.c
index b9d1dd1..94357ba 100644
--- a/src/base/video/remap.c
+++ b/src/base/video/remap.c
@@ -955,7 +955,7 @@ static RectArea remap_mem_1(RemapObject *ro, int offset, int len)
   j2 = (offset + len) % ro->src_scan_len;
 
   /* make sure it's all visible */
-  if(i2 >= ro->src_width) i1++, i2 = 0, offset = i1 * ro->src_scan_len;
+  if(i2 >= ro->src_width) i2 = 0, offset = i1 * ro->src_scan_len;
   if(i1 >= ro->src_height || i1 > j1) return ra;
   if(j2 >= ro->src_width) j1++, j2 = 0;
   if(j1 >= ro->src_height) j1 = ro->src_height, j2 = 0;
@@ -1037,7 +1037,7 @@ static RectArea remap_mem_1(RemapObject *ro, int offset, int len)
     ro->src_y1 = j1;
     if(j2) ro->src_y1++;
     ro->dst_y0 = bre_d_0(i1, ro->src_height, ro->dst_height);
-    ro->dst_y1 = bre_d_0(j1, ro->src_height, ro->dst_height);
+    ro->dst_y1 = bre_d_0(ro->src_y1, ro->src_height, ro->dst_height);
     ro->dst_offset = ro->dst_y0 * ro->dst_scan_len;
     ra.y = ro->dst_y0;
     ra.height = ro->dst_y1 - ro->dst_y0;
```
But nevertheless, I've still have the feeling that this is overcomplicated.

If you want a log, I have one at my other machine which I could provide.

**Comment by bolle732** [comment]
Cleaner in my eyes would be:
```
diff --git a/src/base/video/remap.c b/src/base/video/remap.c
index b9d1dd1..cafd126 100644
--- a/src/base/video/remap.c
+++ b/src/base/video/remap.c
@@ -955,7 +955,7 @@ static RectArea remap_mem_1(RemapObject *ro, int offset, int len)
   j2 = (offset + len) % ro->src_scan_len;
 
   /* make sure it's all visible */
-  if(i2 >= ro->src_width) i1++, i2 = 0, offset = i1 * ro->src_scan_len;
+  if(i2 >= ro->src_width) i2 = 0, offset = i1 * ro->src_scan_len;
   if(i1 >= ro->src_height || i1 > j1) return ra;
   if(j2 >= ro->src_width) j1++, j2 = 0;
   if(j1 >= ro->src_height) j1 = ro->src_height, j2 = 0;
@@ -1034,8 +1034,8 @@ static RectArea remap_mem_1(RemapObject *ro, int offset, int len)
     ro->src_x1 = ro->src_width;
     ro->dst_x1 = ro->dst_width;
     ro->src_y0 = i1;
+    if(j2) j1++;
     ro->src_y1 = j1;
-    if(j2) ro->src_y1++;
     ro->dst_y0 = bre_d_0(i1, ro->src_height, ro->dst_height);
     ro->dst_y1 = bre_d_0(j1, ro->src_height, ro->dst_height);
     ro->dst_offset = ro->dst_y0 * ro->dst_scan_len;
```
This could eventually be adapted to the else clause too... but there it gets overwritten by 'ro->src_height'. Other strange thing.

**Comment by stsp** [comment]
Yes, quite small.
But we need to understand this change better.
For example can the problem be in some rounding
errors of bre_d_0()?

**Comment by bolle732** [comment]
What is the purpose of the 'bre_*' function ? I couldn't find any hint in the source.

**Comment by bolle732** [comment]
Here is an excerpt from my debug output without a patch applied:

===== remap_mem_1 =====
DBG - remap_mem_1: offset=28672, len=45056
DBG - remap_mem_1: remap_func_flags=18, src_scan_len=2560, dst_scan_len=5120
DBG - remap_mem_1: src_width=1280, src_height=854, dst_width=1280, dst_height=854
DBG - remap_mem_1: i1 (y0)=11, i2 (x0)=512, j1 (y1)=28, j2 (x1)=2048
DBG - remap_mem_1: i1 (y0)=11, i2 (x0)=512, j1 (y1)=29, j2 (x1)=0
DBG - remap_mem_1: offset=28672
DBG - remap_mem_1: LINES
DBG - remap_mem_1: i1 (y0)=11, i2 (x0)=512, j1 (y1)=29, j2 (x1)=0
DBG - remap_mem_1: src_x0=0, src_y0=11, src_x1=1280, src_y1=29
DBG - remap_mem_1: dst_x0=0, dst_y0=11, dst_x1=1280, dst_y1=29
DBG - remap_mem_1: ra.x=0, ra.y=11, ra.width=1280, ra.height=18
===== remap_mem_1 =====
DBG - remap_mem_1: offset=32768, len=45056
DBG - remap_mem_1: remap_func_flags=18, src_scan_len=2560, dst_scan_len=5120
DBG - remap_mem_1: src_width=1280, src_height=854, dst_width=1280, dst_height=854
DBG - remap_mem_1: i1 (y0)=12, i2 (x0)=2048, j1 (y1)=30, j2 (x1)=1024
DBG - remap_mem_1: i1 (y0)=13, **i2 (x0)=0**, j1 (y1)=30, j2 (x1)=1024
DBG - remap_mem_1: offset=33280
DBG - remap_mem_1: LINES
DBG - remap_mem_1: i1 (y0)=13, i2 (x0)=0, j1 (y1)=30, j2 (x1)=1024
DBG - remap_mem_1: src_x0=0, src_y0=13, src_x1=1280, **src_y1=31**
DBG - remap_mem_1: dst_x0=0, dst_y0=13, dst_x1=1280, **dst_y1=30**
DBG - remap_mem_1: ra.x=0, ra.y=13, ra.width=1280, ra.height=17

_@ line 934_
===== remap_mem_1 =====
DBG - remap_mem_1: offset=32768, len=45056
DBG - remap_mem_1: remap_func_flags=18, src_scan_len=2560, dst_scan_len=5120
DBG - remap_mem_1: src_width=1280, src_height=854, dst_width=1280, dst_height=854
_@ line 956_
DBG - remap_mem_1: i1 (y0)=12, i2 (x0)=2048, j1 (y1)=30, j2 (x1)=1024
_@ line 962_
DBG - remap_mem_1: i1 (y0)=13, **i2 (x0)=0**, j1 (y1)=30, j2 (x1)=1024
DBG - remap_mem_1: offset=33280
_@ after line 1031 in else if for REF_REMAP_LINES_
DBG - remap_mem_1: LINES
_@ after 1047_
DBG - remap_mem_1: i1 (y0)=13, i2 (x0)=0, j1 (y1)=30, j2 (x1)=1024
DBG - remap_mem_1: src_x0=0, src_y0=13, src_x1=1280, **src_y1=31**
DBG - remap_mem_1: dst_x0=0, dst_y0=13, dst_x1=1280, **dst_y1=30**
_@ line 1065_
DBG - remap_mem_1: ra.x=0, ra.y=13, ra.width=1280, ra.height=17

What can be seen is, if `i2` is set to `0` at line 958 of the 'remap.c', the 'src_y1' and 'dst_y1' of the remapper object aren't set to the same value as I would expect in our case without scaling.

**Comment by stsp** [comment]
> aren't set to the same value as I would expect in our case without scaling.

This is a bug, yes, and it should be fixed by the
change I proposed above. But you say that change
alone doesn't fix the problem. Yet you need to take
it into account when doing any logging.

A side note: what is 854? Do we have a mode 1280x854?

**Comment by bolle732** [comment]
> This is a bug, yes, and it should be fixed by the
change I proposed above. But you say that change
alone doesn't fix the problem. Yet you need to take
it into account when doing any logging.

Do you mean this proposed change:

> Hmm, interesting. In fact, RFF_REMAP_LINES looks
like the most bug-free case here, and is quite simple too.
Can you try to replace at line 1040 j1 with ro->src_y1?

Or did you meant another one ? Sorry, this issue has grown too long...

> A side note: what is 854? Do we have a mode 1280x854?

Yes, it's 1280x854 with 16bpp colour together with the '$_X_vesamode = "1280,854" line... I took this mode, as it produces the artefacts and I know of some users running this mode with DOSEmu(2). But the artefacts occurs in most other standard modes too (ie. 1024x768 with 24bpp).

**Comment by stsp** [comment]
> Do you mean this proposed change:

Yes.

**Comment by bolle732** [comment]
Ok, with that change, the src and dst are equal in case of `i2` is `0`:

===== remap_mem_1 =====
DBG - remap_mem_1: offset=0, len=45056
DBG - remap_mem_1: remap_func_flags=18, src_scan_len=2560, dst_scan_len=5120
DBG - remap_mem_1: src_width=1280, src_height=854, dst_width=1280, dst_height=854
DBG - remap_mem_1: i1 (y0)=0, i2 (x0)=0, j1 (y1)=17, j2 (x1)=1536
DBG - remap_mem_1: i1 (y0)=0, i2 (x0)=0, j1 (y1)=18, j2 (x1)=0
DBG - remap_mem_1: offset=0
DBG - remap_mem_1: LINES
DBG - remap_mem_1: i1 (y0)=0, i2 (x0)=0, j1 (y1)=18, j2 (x1)=0
DBG - remap_mem_1: src_x0=0, src_y0=0, src_x1=1280, src_y1=18
DBG - remap_mem_1: dst_x0=0, dst_y0=0, dst_x1=1280, dst_y1=18
DBG - remap_mem_1: ra.x=0, ra.y=0, ra.width=1280, ra.height=18
===== remap_mem_1 =====
DBG - remap_mem_1: offset=4096, len=45056
DBG - remap_mem_1: remap_func_flags=18, src_scan_len=2560, dst_scan_len=5120
DBG - remap_mem_1: src_width=1280, src_height=854, dst_width=1280, dst_height=854
DBG - remap_mem_1: i1 (y0)=1, i2 (x0)=1536, j1 (y1)=19, j2 (x1)=512
DBG - remap_mem_1: i1 (y0)=2, **i2 (x0)=0**, j1 (y1)=19, j2 (x1)=512
DBG - remap_mem_1: offset=5120
DBG - remap_mem_1: LINES
DBG - remap_mem_1: i1 (y0)=2, i2 (x0)=0, j1 (y1)=19, j2 (x1)=512
DBG - remap_mem_1: src_x0=0, src_y0=2, src_x1=1280, **src_y1=20**
DBG - remap_mem_1: dst_x0=0, dst_y0=2, dst_x1=1280, **dst_y1=20**
DBG - remap_mem_1: ra.x=0, ra.y=2, ra.width=1280, ra.height=18

But still, the artefacts occurs.

**Comment by stsp** [comment]
So my hopes that the problem is localized,
have vanished. :( But at least we can fix a
few obvious bugs, and you have a work-around.

**Comment by bolle732** [comment]
Hm. Not sure how I should understand "you have a workaround"... 

In my eyes, the workaround we came up won't hurt anything.

**Comment by bolle732** [comment]
Ok, this is really good news! Even the change set is much larger.
Thanks a lot !

**Comment by stsp** [comment]
Yes, almost every line in this func had a bug.
